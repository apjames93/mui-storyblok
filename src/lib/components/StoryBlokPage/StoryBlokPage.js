import React, { Component } from 'react';
import { MuiThemeProvider, createMuiTheme } from '@material-ui/core/styles';
import CssBaseline from '@material-ui/core/CssBaseline';
import PropTypes from 'prop-types';
import Grid from '@material-ui/core/Grid';
import MuiCircularProgress from '../MuiCircularProgress/MuiCircularProgress';
import Storyblok from '../../utils/Storyblok';
import styles from './StoryBlokPage.module.scss';
import downloadSbAsset from '../../utils/json-sb-asset';
import { renderComponents } from '../../utils/customComponents';

import PageGrid from '../PageGrid/PageGrid';

export class StoryBlokPage extends Component {
  state = {
    story: [],
    loading: true,
    error: '',
    muiTheme: {},
  };

  components = {
    PageGrid,
  }

  async componentDidMount() {
    this.appendStoryblokBridgeScript(this.props.accessToken);
    await this.getPage();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.theme !== this.props.theme && this.props.useObjectTheme) {
      this.setpropTheme(this.props.theme);
    }
  }

  appendStoryblokBridgeScript = (accessToken) => {
    if (process.env.REACT_APP_ENV !== 'production') {
      const existingScript = document.getElementById('storyblokBridge');
      if (!existingScript) {
        const script = document.createElement('script');
        script.src = `//app.storyblok.com/f/storyblok-latest.js?t=${accessToken}`;
        script.id = 'storyblokBridge';
        document.body.appendChild(script);
        script.onload = () => {
          // Initialize the Storyblok JS Bridge
          window.storyblok.init({ accessToken });

          // Ping the Visual Editor and enter Editmode manually
          window.storyblok.pingEditor(() => {
            if (window.storyblok.inEditor) {
              window.storyblok.enterEditmode();

              window.storyblok.on(['change', 'saved'], () => {
                window.location.reload(true);
              });
            }
          });
        };
      }
    }
  }

  setpropTheme = (theme) => {
    const muiTheme = createMuiTheme(theme);
    this.setState({ muiTheme });
  }

  getUrlTheme = async (urlTheme, theme) => {
    let res;
    try {
      res = await downloadSbAsset(urlTheme);
      return createMuiTheme(res);
    } catch (err) {
      console.warn(`issue getting theme from ${urlTheme} using to ${theme ? 'theme passed by props' : 'mui default theme'}  `);
      return createMuiTheme(theme);
    }
  };

  pickTheme = async (urlTheme, theme) => {
    let muiTheme;
    if (urlTheme && !this.props.useObjectTheme) {
      muiTheme = await this.getUrlTheme(urlTheme);
    } else {
      muiTheme = createMuiTheme(theme);
    }
    return muiTheme;
  };

  getPage = async () => {
    try {
      let route;
      if (window.location.pathname === '/') {
        route = window.location.pathname === '/' ? this.props.dynamicSlug : window.location.pathname.slice(1);
      } else if (this.props.staticSlug) {
        route = this.props.staticSlug;
      } else {
        route = window.location.pathname.slice(1);
      }
      const story = await Storyblok.get(route, this.props.accessToken, this.props.version);
      const muiTheme = await this.pickTheme(story[1], this.props.theme);
      this.setState({ muiTheme });

      this.setStory(story[0]);
    } catch (err) {
      await this.pageNotFound();
    }
  };

  pageNotFound = async () => {
    try {
      const story = await Storyblok.get('page-not-found', this.props.accessToken, this.props.version);
      this.setStory(story);
    } catch (err) {
      this.setState({
        error: 'Unable to fetch page content please refresh and try again or check to make sure you are at the right url',
      });
    }
  };

  setStory = (story) => {
    // scroll to top of page when setting a new story
    document.body.scrollTop = 0; // For Safari
    document.documentElement.scrollTop = 0; // For Chrome, Firefox, IE and Opera
    this.setState({ loading: false, story });
  };

  render() {
    return (
      <Grid
        container
        direction="row"
        justify="center"
        alignItems="center"
        styles={this.props.pageStyles}
      >
        {this.state.loading && !this.state.error && (
          <MuiCircularProgress />
        )}
        {this.state.error && <span style={{ color: 'red' }}>{this.state.error}</span>}
        {!this.state.loading && (
        <MuiThemeProvider theme={this.state.muiTheme}>
          <CssBaseline />
          <div className={styles.container}>
            {this.state.story
                && this.state.story
                  .map((item, index) => renderComponents(this.components, item, index))}
          </div>
        </MuiThemeProvider>
        )}
      </Grid>
    );
  }
}

export default StoryBlokPage;

StoryBlokPage.propTypes = {
  /** If Static Slug is provided storyblok will only load the page provided */
  staticSlug: PropTypes.string,
  /** Override for default route page */
  dynamicSlug: PropTypes.string,
  version: PropTypes.string.isRequired,
  /** acess key from storyblok you can make them in storyblok settings */
  accessToken: PropTypes.string.isRequired,
  /** theme for mui */
  theme: PropTypes.shape(),
  /**
   * override url theme to edit in mui-theme tool
   */
  useObjectTheme: PropTypes.bool,
  pageStyles: PropTypes.shape({}),
};

StoryBlokPage.defaultProps = {
  theme: {},
  useObjectTheme: false,
  dynamicSlug: 'page-welcome',
  staticSlug: '',
  pageStyles: {},
};
