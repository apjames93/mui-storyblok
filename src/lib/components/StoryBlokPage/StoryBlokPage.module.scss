.container {
  width: 100%;
  text-align: center;
  overflow-x: hidden;
  overflow-y: hidden;
}
