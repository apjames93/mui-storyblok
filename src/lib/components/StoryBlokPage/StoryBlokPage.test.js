import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import moxios from 'moxios';
import Storyblok from '../../utils/Storyblok';
import { StoryBlokPage } from './StoryBlokPage';

function setup(useObjectTheme = false, theme = {}) {
  const props = {
    useObjectTheme,
    theme,
    accessToken: 'accessToken',
    version: 'draft',
    location: {
      pathname: 'page-welcome',
    },
  };
  const comp = shallow(<StoryBlokPage {...props} />);
  return { comp, props };
}

describe('getUrlTheme', () => {
  it('getUrlTheme return theme obj', async () => {
    const { comp } = setup();
    const theme = await comp.instance().getUrlTheme('', { cool: 'thing' });
    expect(theme.cool).toEqual('thing');
  });

  it('getUrlTheme return theme url', async () => {
    const { comp } = setup();
    global.fetch = jest.fn(() => Promise.resolve({
      json: () => Promise.resolve({ cool: 'thing' }),
    }));
    const theme = await comp.instance().getUrlTheme('cool.com', {});
    expect(theme.cool).toEqual('thing');
  });
});

describe('pickTheme', () => {
  it('pickTheme return theme url', async () => {
    const { comp } = setup();
    global.fetch = jest.fn(() => Promise.resolve({
      json: () => Promise.resolve({ cool: 'thing' }),
    }));
    const theme = await comp.instance().pickTheme('cool.com', {});
    expect(theme.cool).toEqual('thing');
  });
});

describe('<StoryBlokPage />', () => {
  beforeEach(() => {
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it('renders', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  describe('lifecycle', () => {
    it('componentDidMount calls setStory', async () => {
      const { comp } = setup();
      Storyblok.get = jest.fn(() => ['value']);

      await comp.instance().componentDidMount();
      expect(Storyblok.get).toBeCalled();
      expect(comp.instance().state.story).toEqual('value');
      expect(document.body.scrollTop).toEqual(0);
      expect(document.documentElement.scrollTop).toEqual(0);
    });

    it('componentDidMount calls pageNotFound set page-not-found', async () => {
      const { comp } = setup();
      Storyblok.get = jest.fn(() => {
        throw new Error('woooo');
      });
      await comp.instance().componentDidMount();
      expect(Storyblok.get).toBeCalled();
      Storyblok.get = jest.fn(() => ['value']);
      expect(document.body.scrollTop).toEqual(0);
      expect(document.documentElement.scrollTop).toEqual(0);
      expect(comp.instance().state.story).toEqual('value');
    });

    it('componentDidMount error state', async () => {
      const { comp } = setup();
      Storyblok.get = jest.fn(() => {
        throw new Error('woooo');
      });
      await comp.instance().componentDidMount();

      expect(document.body.scrollTop).toEqual(0);
      expect(document.documentElement.scrollTop).toEqual(0);
      expect(comp.instance().state.error).toEqual('Unable to fetch page content please refresh and try again or check to make sure you are at the right url');
    });


    it('componentDidUpdate', async () => {
      const theme = {
        new: 'theme',
      };

      const { comp } = setup(true, theme);
      expect(comp.instance().state.muiTheme).toEqual({});
      comp.instance().componentDidUpdate({ theme: { old: 'theme' } });
      expect(comp.instance().state.muiTheme.new).toEqual('theme');
    });
  });

  test('snapshot', () => {
    const { props } = setup();
    const tree = renderer.create(<StoryBlokPage {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
