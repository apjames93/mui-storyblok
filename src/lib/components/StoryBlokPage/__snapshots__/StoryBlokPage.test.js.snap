// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StoryBlokPage /> snapshot 1`] = `
<div
  style={Object {}}
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <span
      style={
        Object {
          "color": "red",
        }
      }
    >
      Unable to fetch page content please refresh and try again or check to make sure you are at the right url
    </span>
  </div>
</div>
`;
